fms_core/fzy.py

```diff
@@ -0,0 +1,207 @@
+
+
+def precompute_bonus(haystack):
+    #  Which positions are beginning of words
+    m = len(haystack)
+    match_bonus = [0] * m
+
+    last_ch = '/'
+
+    for i in range(0, m):
+        ch = haystack[i]
+
+        if last_ch == '/':
+            match_bonus[i] = SCORE_MATCH_SLASH
+        elif last_ch == '-' or last_ch == '_' or last_ch == ' ':
+            match_bonus[i] = SCORE_MATCH_WORD
+        elif last_ch == '.':
+            match_bonus[i] = SCORE_MATCH_DOT
+        elif islower(last_ch) and isupper(ch):
+            match_bonus[i] = SCORE_MATCH_CAPITAL
+        else:
+            match_bonus[i] = 0
+
+        last_ch = ch
+
+    return match_bonus
+
+
+def compute(needle, haystack, D, M):
+    n = len(needle)
+    m = len(haystack)
+
+    lower_needle = needle.lower()
+    lower_haystack = haystack.lower()
+
+    match_bonus = precompute_bonus(haystack)
+
+    #
+    # D[][] Stores the best score for this position ending with a match.
+    # M[][] Stores the best possible score at this position.
+    #
+
+    for i in range(0, n):
+        D[i] = [0] * m
+        M[i] = [0] * m
+
+        prev_score = SCORE_MIN
+        gap_score = SCORE_GAP_TRAILING if i == n - 1 else SCORE_GAP_INNER
+
+        for j in range(0, m):
+            if lower_needle[i] == lower_haystack[j]:
+                score = SCORE_MIN
+
+                if not i:
+                    score = (j * SCORE_GAP_LEADING) + match_bonus[j]
+                elif j:  # i > 0 and j > 0*/
+                    score = max([
+                        M[i - 1][j - 1] + match_bonus[j],
+
+                        # consecutive match, doesn't stack with match_bonus
+                        D[i - 1][j - 1] + SCORE_MATCH_CONSECUTIVE])
+
+                D[i][j] = score
+                M[i][j] = prev_score = max([score, prev_score + gap_score])
+            else:
+                D[i][j] = SCORE_MIN
+                M[i][j] = prev_score = prev_score + gap_score
+
+
+def score(needle, haystack):
+    n = len(needle)
+    m = len(haystack)
+
+    if not n or not m:
+        return SCORE_MIN
+
+    #  if n == m:
+        #  #
+        #  # Since this method can only be called with a haystack which
+        #  # matches needle. If len(the)s of the strings are equal the
+        #  # strings themselves must also be equal (ignoring case).
+        #  #
+        #  return SCORE_MAX
+
+    if m > 1024:
+        #
+        # Unreasonably large candidate: return no score
+        # If it is a valid match it will still be returned, it will
+        # just be ranked below any reasonably sized candidates
+        #
+        return SCORE_MIN
+
+    D = [0] * n
+    M = [0] * n
+
+    compute(needle, haystack, D, M)
+
+    return M[n - 1][m - 1]
+
+
+def positions(needle, haystack):
+    n = len(needle)
+    m = len(haystack)
+
+    positions = []
+
+    if not n or not m:
+        return positions
+
+    if n == m:
+        for i in range(0, n):
+            positions[i] = i
```

Comment:
wouldn't this cause index out of bounds errors?

Reply:
also, if it's just assigning the values from range() to the list, it could be
```python
positions = list(range(n))
```

Reply:
indeed it does, my python isn't good TBH